### PR TITLE
[WIP] Draft version of GraphClient

### PR DESCRIPTION
**Pull request by sav007**
**DO NOT MERGE**

### CODE REVIEW

File: MobileBuy/buy/src/main/java/com/shopify/buy3/GraphCall.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  void cancel();
+
+  @NonNull GraphCall<T> clone();
+
+  @NonNull T execute() throws IOException, SchemaViolationError;
+
+  @NonNull GraphCall<T> enqueue(@NonNull Callback<T> callback);
+
+  interface Callback<T extends AbstractResponse<T>> {
+    void onResponse(@NonNull T response);
```
Comment (krisorr) [inline MobileBuy/buy/src/main/java/com/shopify/buy3/GraphCall.java]: add a new line?

File: MobileBuy/buy/src/main/java/com/shopify/buy3/GraphClient.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+      return this;
+    }
+
+    /**
+     * Sets custom OkHttp httpInterceptors
+     *
+     * @param httpInterceptors Interceptors to add to the OkHttp client.
+     * @return a {@link BuyClientBuilder}
+     */
+    public Builder httpInterceptors(final Interceptor... httpInterceptors) {
```
Comment (krisorr) [inline MobileBuy/buy/src/main/java/com/shopify/buy3/GraphClient.java]: hmm, I guess we could still hook into these to do a simple http cache even with this client

File: MobileBuy/buy/src/main/java/com/shopify/buy3/RealGraphCall.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    }
+    httpCall = httpCall();
+    Response response = httpCall.execute();
+    TopLevelResponse topLevelResponse = parse(response);
+    return responseConverter.convert(topLevelResponse);
+  }
+
+  @NonNull @Override public GraphCall<T> enqueue(@NonNull final Callback<T> callback) {
+    synchronized (this) {
+      if (executed) throw new IllegalStateException("Already Executed");
```
Comment (krisorr) [inline MobileBuy/buy/src/main/java/com/shopify/buy3/RealGraphCall.java]: do we want to throw and exception or just drop it?

File: MobileBuy/buy/src/main/java/com/shopify/buy3/GraphClient.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+  @Override public GraphCall<APISchema.Mutation> mutateGraph(final APISchema.MutationQuery query) {
+    return new RealGraphCall<>(query, serverUrl, httpCallFactory, response -> new APISchema.Mutation(response.getData()));
+  }
+
+  private OkHttpClient buildHttpCallFactory(final String apiKey, final String applicationName, final long httpConnectionTimeoutMs,
+                                            final long httpReadWriteTimeoutMs, final Interceptor[] httpInterceptors) {
+    final Interceptor requestInterceptor = chain -> {
+      final Request original = chain.request();
+      final Request.Builder builder = original.newBuilder().method(original.method(), original.body());
```
Comment (krisorr) [inline MobileBuy/buy/src/main/java/com/shopify/buy3/GraphClient.java]: can we get the server updated so we can pass the auth as part of the payload instead?  I know there was talk about at least doing that for customer credentials.

Comment (sav007) [inline MobileBuy/buy/src/main/java/com/shopify/buy3/GraphClient.java]: We are trying to find out with server side, if it's possible

File: MobileBuy/buy/src/main/java/com/shopify/buy3/RealGraphCall.java (hunk 3 above)
Comment (sav007) [inline MobileBuy/buy/src/main/java/com/shopify/buy3/RealGraphCall.java]: This is precondition exception, as we make sure that we don't run queries multiple times. This logic is similar to what behind `okhttp3.RealCall#enqueue`:
```
@Override public void enqueue(Callback responseCallback) {
    synchronized (this) {
      if (executed) throw new IllegalStateException("Already Executed");
      executed = true;
    }
    captureCallStackTrace();
    client.dispatcher().enqueue(new AsyncCall(responseCallback));
  }
```

**Comment by krisorr** [comment]
LGTM, only the one question about the credentials, and that could be addressed in a later PR.

**Comment by krisorr** [comment]
Still LGTM :)

**Pull request merged**